File: Android/app/src/androidTest/java/jp/co/riso/android/util/FileUtilsTest.java

package jp.co.riso.android.util;

import android.content.res.AssetManager;
import android.net.Uri;
import android.test.ActivityInstrumentationTestCase2;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;

import jp.co.riso.smartdeviceapp.SmartDeviceApp;
import jp.co.riso.smartdeviceapp.view.MainActivity;

public class FileUtilsTest extends ActivityInstrumentationTestCase2<MainActivity> {
    final private static String TEST_SRC_FILE = "PDF-squarish.pdf";
    final private static String TEST_DST_FILE = "sample_dst.pdf";

    private File mSrcFile = null;
    private File mDstFile = null;

    public FileUtilsTest() {
        super(MainActivity.class);
    }

    public FileUtilsTest(Class<MainActivity> activityClass) {
        super(activityClass);
    }

    protected void setUp() throws Exception {
        super.setUp();
    }

    protected void tearDown() throws Exception {
        super.tearDown();
    }

    // ================================================================================
    // Tests - constructors
    // ================================================================================

    public void testConstructor() {
        FileUtils utils = new FileUtils();
        assertNotNull(utils);
    }

    // ================================================================================
    // Tests - copy
    // ================================================================================

    public void testCopy_ValidParams() {
        try {
            String pdfPath = getAssetPath(TEST_SRC_FILE);

            mSrcFile = new File(pdfPath);
            if (mSrcFile == null) {
                fail();
            }

            mDstFile = new File(SmartDeviceApp.getAppContext().getExternalFilesDir("pdfs") + "/"
                    + TEST_DST_FILE);
            if (mDstFile == null) {
                fail();
            }

            FileUtils.copy(mSrcFile, mDstFile);
        } catch (Exception e) {
            fail();
        }
    }

    public void testCopy_NullSrc() {
        try {
            FileUtils.copy((InputStream) null, mDstFile);
        } catch (Exception e) {
            fail();
        }
    }

    public void testCopy_NullDst() {
        try {
            FileUtils.copy(mSrcFile, null);
        } catch (Exception e) {
            fail();
        }
    }

    // ================================================================================
    // Tests - get filename
    // ================================================================================

    public void testGetFilename() {
        try {
            String pdfPath = getAssetPath(TEST_SRC_FILE);

            mSrcFile = new File(pdfPath);
            if (mSrcFile == null) {
                fail();
            }

            assertEquals(FileUtils.getFileName(SmartDeviceApp.getAppContext(), Uri.fromFile(mSrcFile), true), TEST_SRC_FILE);
        } catch (Exception e) {
            fail();
        }
    }

    // ================================================================================
    // Tests - get filesize
    // ================================================================================

    public void testGetFileSize() {
        try {
            String pdfPath = getAssetPath(TEST_SRC_FILE);
            mSrcFile = new File(pdfPath);
            assertEquals(FileUtils.getFileSize(SmartDeviceApp.getAppContext(), Uri.fromFile(mSrcFile)), mSrcFile.length());
        } catch (Exception e) {
            fail();
        }
    }

    // ================================================================================
    // Tests - get mimetype
    // ================================================================================

    public void testGetMimeType() {
        try {
            String pdfPath = getAssetPath(TEST_SRC_FILE);
            mSrcFile = new File(pdfPath);
            assertEquals(FileUtils.getMimeType(SmartDeviceApp.getAppContext(), Uri.fromFile(mSrcFile)), "application/pdf");
        } catch (Exception e) {
            fail();
        }
    }

    // ================================================================================
    // Private
    // ================================================================================

    private String getAssetPath(String filename) {
        File f = new File(getActivity().getCacheDir() + "/" + filename);
        AssetManager assetManager = getInstrumentation().getContext().getAssets();

        try {
            InputStream is = assetManager.open(filename);
            int size = is.available();
            byte[] buffer = new byte[size];
            is.read(buffer);
            is.close();
            
            FileOutputStream fos = new FileOutputStream(f);
            fos.write(buffer);
            fos.close();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
        
        return f.getPath();
    }

}

File: Android/app/src/androidTest/kotlin/jp/co/riso/android/util/FileUtilsTest.kt
package jp.co.riso.android.util

import android.net.Uri
import androidx.test.platform.app.InstrumentationRegistry
import jp.co.riso.android.util.FileUtils.copy
import jp.co.riso.android.util.FileUtils.delete
import jp.co.riso.android.util.FileUtils.getFileName
import jp.co.riso.android.util.FileUtils.getFileSize
import jp.co.riso.android.util.FileUtils.getMimeType
import jp.co.riso.smartdeviceapp.SmartDeviceApp
import jp.co.riso.smartdeviceapp.view.BaseActivityTestUtil
import junit.framework.TestCase
import org.junit.Test
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.InputStream

class FileUtilsTest : BaseActivityTestUtil() {
    private var _srcFile: File? = null
    private var _srcInputStream: InputStream? = null
    private var _dstFile: File? = null
    
    // ================================================================================
    // Tests - constructors
    // ================================================================================
    @Test
    fun testConstructor() {
        val utils = FileUtils
        TestCase.assertNotNull(utils)
    }

    // ================================================================================
    // Tests - copy
    // ================================================================================
    @Test
    fun testCopy_ValidParams() {
        try {
            val pdfPath = getAssetPath(TEST_SRC_FILE)
            _srcFile = File(pdfPath)
            if (_srcFile == null) {
                TestCase.fail()
            }
            _dstFile = File(
                SmartDeviceApp.appContext!!.getExternalFilesDir("pdfs").toString() + "/"
                        + TEST_DST_FILE
            )
            if (_dstFile == null) {
                TestCase.fail()
            }
            copy(_srcFile, _dstFile)
        } catch (e: Exception) {
            TestCase.fail()
        }
    }

    @Test
    fun testCopy_ValidParams2() {
        try {
            val pdfPath = getAssetPath(TEST_SRC_FILE)
            _srcInputStream = FileInputStream(pdfPath)
            if (_srcInputStream == null) {
                TestCase.fail()
            }
            _dstFile = File(
                SmartDeviceApp.appContext!!.getExternalFilesDir("pdfs").toString() + "/"
                        + TEST_DST_FILE
            )
            if (_dstFile == null) {
                TestCase.fail()
            }
            copy(_srcInputStream, _dstFile)
        } catch (e: Exception) {
            TestCase.fail()
        }
    }

    @Test
    fun testCopy_NullSrc() {
        try {
            _dstFile = File(
                SmartDeviceApp.appContext!!.getExternalFilesDir("pdfs").toString() + "/"
                        + TEST_DST_FILE
            )
            copy(_srcInputStream, _dstFile)
            copy(_srcFile, _dstFile)
        } catch (e: Exception) {
            TestCase.fail()
        }
    }

    @Test
    fun testCopy_NullDst() {
        try {
            val pdfPath = getAssetPath(TEST_SRC_FILE)
            _srcInputStream = FileInputStream(pdfPath)
            _srcFile = File(pdfPath)

            copy(_srcInputStream, _dstFile)
            copy(_srcFile, _dstFile)
        } catch (e: Exception) {
            TestCase.fail()
        }
    }

    @Test
    fun testCopy_NullParameters() {
        try {
            copy(_srcInputStream, _dstFile)
            copy(_srcFile, _dstFile)
        } catch (e: Exception) {
            TestCase.fail()
        }
    }

    // ================================================================================
    // Tests - get filename
    // ================================================================================
    @Test
    fun testGetFilename() {
        try {
            val pdfPath = getAssetPath(TEST_SRC_FILE)
            _srcFile = File(pdfPath)
            if (_srcFile == null) {
                TestCase.fail()
            }
            TestCase.assertEquals(
                getFileName(
                    SmartDeviceApp.appContext,
                    Uri.fromFile(_srcFile),
                    true
                ), TEST_SRC_FILE
            )
        } catch (e: Exception) {
            TestCase.fail()
        }
    }

    // ================================================================================
    // Tests - get filesize
    // ================================================================================
    @Test
    fun testGetFileSize() {
        try {
            val pdfPath = getAssetPath(TEST_SRC_FILE)
            _srcFile = File(pdfPath)
            TestCase.assertEquals(
                getFileSize(
                    SmartDeviceApp.appContext,
                    Uri.fromFile(_srcFile)
                ).toLong(), _srcFile!!.length()
            )
        } catch (e: Exception) {
            TestCase.fail()
        }
    }

    // ================================================================================
    // Tests - get mimetype
    // ================================================================================
    @Test
    fun testGetMimeType() {
        try {
            val pdfPath = getAssetPath(TEST_SRC_FILE)
            _srcFile = File(pdfPath)
            TestCase.assertEquals(
                getMimeType(
                    SmartDeviceApp.appContext,
                    Uri.fromFile(_srcFile)
                ), "application/pdf"
            )
        } catch (e: Exception) {
            TestCase.fail()
        }
    }

    @Test
    fun testDelete() {
        try {
            val pdfPath = getAssetPath(TEST_SRC_FILE)
            _srcFile = File(pdfPath)

            delete(_srcFile!!)
        } catch (e: Exception) {
            TestCase.fail()
        }
    }

    // ================================================================================
    // Private
    // ================================================================================
    private fun getAssetPath(filename: String): String {
        val f = File(mainActivity!!.cacheDir.toString() + "/" + filename)
        val assetManager = InstrumentationRegistry.getInstrumentation().context.assets
        try {
            val `is` = assetManager.open(filename)
            val size = `is`.available()
            val buffer = ByteArray(size)
            `is`.read(buffer)
            `is`.close()
            val fos = FileOutputStream(f)
            fos.write(buffer)
            fos.close()
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
        return f.path
    }

    companion object {
        private const val TEST_SRC_FILE = "PDF-squarish.pdf"
        private const val TEST_DST_FILE = "sample_dst.pdf"
    }
}